pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,452 @@
+// Package libp2p implements the Mir transport interface using libp2p.
+package libp2p
+
+//go:generate go run ./gen/gen.go
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/logging"
+	mirnet "github.com/filecoin-project/mir/pkg/net"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/messagepb"
+	"github.com/filecoin-project/mir/pkg/types"
+)
+
+const (
+	ProtocolID             = "/mir/0.0.1"
+	maxConnectingTimeout   = 1000 * time.Millisecond
+	maxRetryTimeout        = 2 * time.Second
+	maxRetries             = 20
+	noLoggingErrorAttempts = 2
+	PermanentAddrTTL       = math.MaxInt64 - iota
+)
+
+type TransportMessage struct {
+	Sender  string
+	Payload []byte
+}
+
+var _ mirnet.Transport = &Transport{}
+
+var ErrUnknownNode = errors.New("unknown node")
+var ErrNilStream = errors.New("stream has not been opened")
+
+type connInfo struct {
+	AddrInfo  *peer.AddrInfo
+	Stream    network.Stream
+	IsOpening bool
+}
+
+type Transport struct {
+	host             host.Host
+	ownID            types.NodeID
+	logger           logging.Logger
+	incomingMessages chan *events.EventList
+	connWg           *sync.WaitGroup
+	conns            map[types.NodeID]*connInfo
+	connsLock        sync.RWMutex
+}
+
+func NewTransport(h host.Host, ownID types.NodeID, logger logging.Logger) (*Transport, error) {
+	if logger == nil {
+		logger = logging.ConsoleErrorLogger
+	}
+
+	return &Transport{
+		ownID:            ownID,
+		host:             h,
+		incomingMessages: make(chan *events.EventList),
+		logger:           logger,
+		connWg:           &sync.WaitGroup{},
+		conns:            make(map[types.NodeID]*connInfo),
+	}, nil
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, "starting libp2p transport", "ownID", t.ownID, "listen", t.host.Addrs())
+	t.host.SetStreamHandler(ProtocolID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "stopping libp2p transport", "ownID", t.ownID)
+	defer t.logger.Log(logging.LevelDebug, "stopping libp2p transport finished.", "ownID", t.ownID)
+
+	t.host.RemoveStreamHandler(ProtocolID)
```

Comment:
Does this also close incoming connections?

Reply:
It doesn't close connections directly
https://github.com/multiformats/go-multistream/blob/225d479012e89db2156d325a60b9374713f21769/multistream.go#L148

Reply:
Then I think we'd better close all the streams in `mirHandler`.

Reply:
`mirHandler` is the same as a handler in`http`. I don't see how to implement that.

Reply:
It receives `network.Stream` which has `Close` method. I think we should arrange all loops running in `mirHandler` to call `s.Close` once we call `Stop`.

Reply:
The difference from an HTTP handler is that `mirHandler` has a long-running loop inside. If we stop the transport module, it should also drop the incoming connections, otherwise the corresponding resources cannot be garbage-collected.

Reply:
That shows that it doesn't close connections.

Reply:
Probably, I did not understand you right.

Reply:
I suspect that neither `RemoveStreamHandler` nor `Network().ClosePeer` would close an already established incoming stream connection.

Reply:
Stream and/or connection :)

Reply:
> ClosePeer closes the connection to a given peer

Here I'm not sure it also closes incoming connections.

Reply:
FWIW (and I just skimmed through the discussion), I usually close all streams before closing the host as a sanity check with something like:
```go 
// Unregister handler
...
// Close streams
for _, s := range streams{
   s.Close()
}
// Close host
h.Close()
```

Reply:
If I understood Sergay correctly he suggests closing all streams in the handler. 
Currently, the handler closes the stream on defer.

Reply:
Yes, on `Stop`, I suggest closing all the streams received in `mirHandler` and waiting until all handlers exit.